File: DSA/Java/src/dsa/lib/algo/sort/BucketSort.java
package dsa.lib.algo.sort;

import dsa.lib.utils.Arrays;

class BucketSort {
    private int[] nums;
    private final int size;
    int min, max;

    private int[][] buckets;
    private int[] bucketSizes;
    int bucketCount;
    private final int capacity;

    public BucketSort(int[] nums) {
        this.nums = nums;
        this.size = nums.length;
        this.capacity = 2;
    }

    public BucketSort(int[] nums, int capacity) {
        this.nums = nums;
        this.size = nums.length;
        this.capacity = capacity;
    }

    public void sort() {
        if (this.size < 2) return;

        initFields();
        scatterToEachBucket();
        sortEachBucket();
    }

    private void initFields() {
        initMinMax();
        initBucketCount();
        initBuckets();
        initBucketSizes();
    }

    private void initMinMax() {
        if (this.size == 0) return;

        min = max = nums[0];
        for (final int num : this.nums) {
            if (num < min) min = num;
            else if (num > max) max = num;
        }
    }

    private void initBucketCount() {
        bucketCount = (max - min + 1) / capacity + 1;
    }

    private void initBuckets() {
        buckets = new int[bucketCount][capacity];
    }

    private void initBucketSizes() {
        bucketSizes = new int[bucketCount];
    }

    private void scatterToEachBucket() {
        for (final int num : nums) {
            int index = bucketIndexOf(num);
            int size = bucketSizes[index];

            int cap = buckets[index].length;
            if (size >= cap) resizeBucket(index, (int) (cap * 1.5f));

            buckets[index][size] = num;
            ++bucketSizes[index];
        }
    }

    private int bucketIndexOf(int num) {
        return (num - min) / capacity;
    }

    private void resizeBucket(int index, int newSize) {
        int[] newBucket = new int[newSize];
        int[] oldBucket = buckets[index];
        Arrays.copy(oldBucket, 0, newBucket, 0, oldBucket.length);
        buckets[index] = newBucket;
    }

    private void sortEachBucket() {
        int k = 0;
        for (int i = 0; i < buckets.length; ++i) {
            int[] bucket = buckets[i];
            int size = bucketSizes[i];

            if (size == 0) continue;

            quickSort(bucket, size);
            Arrays.copy(bucket, 0, nums, k, size);
            k += size;
        }
    }

    private void quickSort(int[] nums, int n) {
        new QuickSort().sort(nums, n);
    }
}

File: DSA/Python/src/dsa/lib/algo/sort/bucket_sort.py
from typing import List

from dsa.lib import utils
from dsa.lib.algo.sort.quick_sort import QuickSort


class BucketSort:

    def __init__(self, nums: List[int], capacity: int = 2) -> None:
        self.nums: List[int] = nums
        self.capacity: int = capacity

        self.min: int = 0
        self.max: int = 0
        self.bucketCount: int = 0
        self.buckets: List[List[int]] = [[]]
        self.bucketSizes: List[int] = []

    def sort(self) -> None:
        if len(self.nums) < 2: return

        self.init_fields()
        self.scatterToEachBucket()
        self.sortEachBucket()

    def init_fields(self):
        self.init_minMax()
        self.init_bucketCount()
        self.init_buckets()
        self.init_bucketSizes()

    def init_minMax(self) -> None:
        self.min = self.max = self.nums[0]
        for num in self.nums:
            if num < self.min:
                self.min = num
            elif num > self.max:
                self.max = num

    def init_bucketCount(self) -> None:
        self.bucketCount = (self.max - self.min + 1) // self.capacity + 1

    def init_buckets(self) -> None:
        self.buckets = [[] for _ in range(self.bucketCount)]

    def init_bucketSizes(self) -> None:
        self.bucketSizes = [0] * self.bucketCount

    def scatterToEachBucket(self) -> None:
        for num in self.nums:
            i: int = self.bucketIndexOf(num)
            self.buckets[i].append(num)
            self.bucketSizes[i] += 1

    def bucketIndexOf(self, num: int) -> int:
        return (num - self.min) // self.capacity

    def sortEachBucket(self) -> None:
        k: int = 0
        for bucket in self.buckets:
            self.quick_sort(bucket)
            size: int = len(bucket)
            utils.array_copy(bucket, 0, self.nums, k, size)
            k += size

    @staticmethod
    def quick_sort(nums: List[int]) -> None:
        QuickSort().sort(nums)
